### PR TITLE
Migrate from Rollup to TSC

### PR DESCRIPTION
**Pull request by macalinao**
This allows inspecting sources when consuming the package, and the build is significantly faster.

I still need to test to see if the generated NPM package works. Putting this PR up for tracking purposes.

Other notes:

- the compiled tests seem to be included in the final build, since tests are covered by the current `tsconfig.json`. This should not affect bundle size but build times (which are really fast) can be reduced by using a different tsconfig for tests
- I removed the `baseUrl` from `tsconfig.json` so we don't have to hardcode a path to `node_modules` for any package

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -18,49 +18,26 @@
...
         "jsx": "preserve",
         "stripInternal": true,
-        "baseUrl": ".",
         "rootDirs": [
             "./src",
         ],
         "declaration": true,
         "sourceMap": true,
         "paths": {
-            "react": [
```
Comment (macalinao) [inline tsconfig.json:28]: Removed these references to node_modules to simplify the build

**Comment by macalinao** [comment]
There seems to be an issue where importing the package is causing my tsc to error saying "Cannot find name 'process'." I'm looking into this right now

**Comment by jmeistrich** [comment]
@macalinao I updated the checksize scripts to the new build locations. Interestingly it added 2 bytes to the gzipped core bundle but reduced 77 bytes from the gzipped react bundle.

**Comment by macalinao** [comment]
> @macalinao I updated the checksize scripts to the new build locations. Interestingly it added 2 bytes to the gzipped core bundle but reduced 77 bytes from the gzipped react bundle.

Nice. We should be able to reduce size further by using `tslib`, which I think we should implement in a later PR

**Comment by jmeistrich** [comment]
Is it working in your codebases? In Next.js it seems like it's resolving the root package but not the subfolders like `/persist`. Are you seeing that too?

**Comment by macalinao** [comment]
> Is it working in your codebases? In Next.js it seems like it's resolving the root package but not the subfolders like `/persist`. Are you seeing that too?

I only tested @legendapp/state/react-- will look into subfolders

did you clear the Next.js cache? And are you using the root package.json as the linked package, rather than the one in `dist/`?

**Comment by jmeistrich** [comment]
> I only tested @legendapp/state/react-- will look into subfolders

I'm seeing the error on the /react path too.

> did you clear the Next.js cache? And are you using the root package.json as the linked package, rather than the one in `dist/`?

I did an `rm -rf .next` before running the Next project which afaik clears the cache, or is there a better way to clear the cache? I emptied the `node_modules/@legendapp/state` folder and copied the whole package folder into there.

> Also what Node version are you testing on? I am on Node 18

I'm using Next.js v13.2.4 and I tried on Node 16 and 18. Both of them have issues with loading .ts files, I think because the file named "react" is `react.ts` so it's trying to load that?

image

**Comment by macalinao** [comment]
> > I only tested @legendapp/state/react-- will look into subfolders
> 
> I'm seeing the error on the /react path too.
> 
> > did you clear the Next.js cache? And are you using the root package.json as the linked package, rather than the one in `dist/`?
> 
> I did an `rm -rf .next` before running the Next project which afaik clears the cache, or is there a better way to clear the cache? I emptied the `node_modules/@legendapp/state` folder and copied the whole package folder into there.
> 
> > Also what Node version are you testing on? I am on Node 18
> 
> I'm using Next.js v13.2.4 and I tried on Node 16 and 18. Both of them have issues with loading .ts files, I think because the file named "react" is `react.ts` so it's trying to load that?
> 
> image

Hmm I was having a similar error where Parcel was trying to compile the source but it disappeared after clearing the cache

The `.next/` thing should theoretically work. Did you also make sure the package was built? And cleared `dist/`?  And maybe deleted `node_modules`? (I am using PNP so I don't have a `node_modules` directory)

I'll try running my code again-- maybe I missed something

**Comment by macalinao** [comment]
Just tested it again (this time published to NPM as `@simplyianm/legend-state`) -- react import works but I found an issue with the wildcard imports, which I'll push a fix for shortly

**Comment by jmeistrich** [comment]
Hmm, I changed the page to return null and just do this:

```js
import { observable } from '@simplyianm/legend-state';
import { useComputed } from '@simplyianm/legend-state/react';

const o = observable({ test: 'helloz' });
console.log(o.peek());
console.log(useComputed);

export default (function Clients() {
    return null;
});
```

And it fails with the same error.

image

I think having uncompiled .ts files in the root is problematic? I believe some environments don't respect `exports` in package.json and just use the file structure? I'm not sure what could be the difference between our projects?

**Comment by chakrihacker** [comment]
Another platform that still has issues with exports is repack. I have to fallback to cjs everytime

**Comment by macalinao** [comment]
Hmm not sure what happened with my Next.js setup-- I can take a look when I have time, though it seems I can't get the wildcard exports working on my setups with Parcel and Next.js. I think there are a few solutions here:

1. We can split the package into subpackages compiled with TSC, e.g. `@legendapp/state` and `@legendapp/state-react`, then create a legacy package bundled with Rollup which references the subpackages
2. We can force any package consumers to change their imports, which would require a major bump

Might be more trouble than it's worth-- since the original issue was around not being able to see sources with the existing package setup, I might try tackling that individually.

**Comment by macalinao** [comment]
Hmm not sure what happened with my Next.js setup, but I can take a look again.

To resolve the sources issue, maybe I should try moving all sources into the `src/` directory and seeing if it causes the bundlers to not resolve the TypeScript source files as the main file. Will experiment with this

**Comment by macalinao** [comment]
I think the exports will break on any package consumers not using Node's new module resolution-- will see if there is a way to fix this at least for Next.js

**Comment by jmeistrich** [comment]
In the latest v3 alpha I migrated from rollup to tsup which is significantly faster and should be more correct. So since this PR is now very out of date I’m going to close it. But please let me know if you want to pick up the concept again on the latest version.

**Pull request closed without merging**